src/app/profile/favorites/page.tsx

```diff
@@ -1,14 +1,47 @@
+import FavoriteProductGallery from "@/components/FavoriteProductsGallery";
+import { getFavorites } from "@/model/FavoriteProductsModel";
+import { getProducts } from "@/model/ProductModel";
+import { Product } from "@/services/ProductAPI";
+import getUserServerSide from "@/services/server/UserService";
 import { faHeart } from "@fortawesome/free-solid-svg-icons";
 import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
 
-function Favorites() {
+async function Favorites() {
+  const user = await getUserServerSide();
+  let favoriteProductsIds: string[] = [];
+  let products: Product[] = [];
+
+  if (user !== undefined) {
+    // any user that is not logged in and tries to access a /profile page
+    // will be redirected to the login page. (code in app/profile/layout.tsx)
+    try {
+      let favorites = await getFavorites(user.id);
+      favoriteProductsIds = favorites.map((favorite) => favorite.id);
+    } catch (err) {
+      console.error("Failed to show favorite products", err);
+    }
+  }
+
+  try {
+    products = await getProducts();
+  } catch (err) {
+    console.error("Failed to select products", err);
+  }
```

Comment:
I just realized you copy pasted this code from the `app/page.tsx` BUT we don't actually need it.

Here, on this page, we want ONLY the favorite products. Which you can get via the `getFavorites` function.

Why fetch ALL the Products, and then filter on them? 😆😆😆  Just use `getFavorites` and voila, you're done.